MSC2448: Using BlurHash as a Placeholder for Matrix Media (#2448)
matrix-org/matrix-doc: proposals/2448-blurhash-for-media.md

thumbnail downloads.

## Proposal

### m.room.message

A new optional field is added in `m.room.message`'s content field called
`blurhash`. It is a BlurHash of the original piece of media. Clients could
then render this using [one of the available BlurHash
implementations](https://github.com/woltapp/blurhash).

This would be optionally displayed while the thumbnail of the media is loaded
in parallel.

Example `m.room.message` content field:

```
{
  "body": "image.png",
  "info": {
    "size": 149234,
    "mimetype": "image/png",
    "thumbnail_info": {
      "w": 301,
      "h": 193,
      "mimetype": "image/png",
      "size": 172958
    },
    "w": 301,
    "h": 193,
    "thumbnail_url": "mxc://example.org/abcdefg"
  },
  "msgtype": "m.image",
  "url": "mxc://example.org/abcde",
  "blurhash": "JadR*.7kCMdnj"
}
```

### Inline images

An optional attribute is added to `<img>` tags in messages: `blurhash`, where
the value of the attribute is the blurhash representation of the inline
image.

This would be optionally displayed with the inline image itself is loaded in parallel.

Example `m.room.message.formatted_body`:

```

> "formatted_body": This is awesome <img alt=\"flutterjoy\" title=\"flutterjoy\" height=\"32\" src=\"mxc://matrix.example.org/abc\" blurhash=\"LEHV6nWB2yk8pyo\" />

Reviewer: This isn't part of `img`, so presumably this should be `data-blurhash` ?

> "formatted_body": This is awesome <img alt=\"flutterjoy\" title=\"flutterjoy\" height=\"32\" src=\"mxc://matrix.example.org/abc\" blurhash=\"LEHV6nWB2yk8pyo\" />

Reviewer 2: Do blurhashes include any characters that are unsafe in this context?

Author: It does not produce any `"` characters.

Reviewer 2: Wonderful!

Reviewer 3: Other characters like `?` and `&` should be HTML-encoded in parameters as per HTML spec, too, though

Author: Ah yeah. Is it fine if we just specify that characters should be URL encoded?

Reviewer 3: HTML encoded isn't URL encoded.

Author: Yep, just realized. Have put into the doc for clients to figure out.

Reviewer 3: Soru doesn't think that belongs in this MSC as it is already part of HTML spec to HTML-encode arguments of tags. She just wanted to point this out as the discussion came up.

Author: Fair enough. I think it's helpful for the MSC but not for the spec.

I'll just take it out.

Reviewer 4: The character set that's used is HTML-safe https://github.com/woltapp/blurhash/blob/master/Algorithm.md#base-83  No `<`, `>`, `&`, or `"` (and `?` doesn't need to be escaped in HTML)

Reviewer 3: > (and `?` doesn't need to be escaped in HTML)

Gotcha, remembered that wrong, then.

If you rely on the info endpoint it'd be a good idea to html sanitize it anyways to rule out attack vectors :+1:

```

Note that a BlurHash representation is really only applicable to media, and
as such should only be used in conjunction with the following
`m.room.message` msgtypes:

* `m.image`
* `m.video`

To be clear: This does not replace thumbnails - it will be shown before they
are downloaded.

## Calculating a blurhash

BlurHashes are inserted into `m.room.message` events by the client, however
some clients may not be able to implement the BlurHash library for whatever
reason. In this case, it would be nice to allow the media repository to
calculate the BlurHash of a piece of media for the client, similar to how
thumbnails are calculated by media repositories today.

> The [`/_matrix/media/r0/upload`](https://matrix.org/docs/spec/client_server/r0.6.0#post-matrix-media-r0-upload) endpoint response is modified to include an optional `blurhash` parameter, which the client may use to insert into messages if desired:

Reviewer: Wrap please

> The [`/_matrix/media/r0/upload`](https://matrix.org/docs/spec/client_server/r0.6.0#post-matrix-media-r0-upload) endpoint response is modified to include an optional `blurhash` parameter, which the client may use to insert into messages if desired:

Reviewer: ```suggestion
The [`/_matrix/media/r0/upload`](https://matrix.org/docs/spec/client_server/r0.6.0#post-matrix-media-r0-upload) endpoint response is modified to include an optional `blurhash` parameter, which the client may use to insert into messages if desired:
```
```suggestion
The [`/_matrix/media/r0/upload`](https://matrix.org/docs/spec/client_server/r0.6.0#post-matrix-media-r0-upload) endpoint response is modified to include an optional `blurhash` key, which the client may use to insert into messages if desired:
```

# MSC2448: Using BlurHash in Media Events

[BlurHash](https://blurha.sh) is a compact representation of a placeholder
for an image (or the frame of a video). Currently in Matrix, clients must
display a placeholder image in the message timeline while a piece of media is
loading. Some clients, such as Riot, simply display an empty space.

While thumbnails exist to combat this to some degree, they still need to be
downloaded from a homeserver, which is not instantaneous.

Instead, a BlurHash can be sent inside the `m.room.message` event, which upon
receipt other clients can render for a pretty preview while the actual
thumbnail downloads. They also do not contain any `"` characters, making them
simple to stick inside existing JSON blobs.

## Proposal

### m.room.message

> A new optional field is added in `m.room.message`'s content field called

Reviewer 3: Maybe specify that this only applies to msgtype `m.image` and `m.video`?

Author: I do, though moved it around to make it hopefully clearer.

Reviewer: /me cries in extensible events

`blurhash`. It is a BlurHash of the original piece of media. Clients could
then render this using [one of the available BlurHash
implementations](https://github.com/woltapp/blurhash).

This would be optionally displayed while the thumbnail of the media is loaded
in parallel.

Example `m.room.message` content field:

```
{
  "body": "image.png",
  "info": {
    "size": 149234,
    "mimetype": "image/png",
    "thumbnail_info": {
      "w": 301,
      "h": 193,
      "mimetype": "image/png",

>       "size": 172958

Reviewer 3: your thumbnail is larger in size than your original image :upside_down_face:

    },
    "w": 301,
    "h": 193,
    "thumbnail_url": "mxc://example.org/abcdefg"
  },
  "msgtype": "m.image",
  "url": "mxc://example.org/abcde",
  "blurhash": "JadR*.7kCMdnj"
}
```

### Inline images

An optional attribute is added to `<img>` tags in messages: `data-blurhash`,
where the value of the attribute is the blurhash representation of the inline
image.

This would be optionally displayed with the inline image itself is loaded in parallel.

Example `m.room.message.formatted_body`:

```

> "formatted_body": This is awesome <img alt=\"flutterjoy\" title=\"flutterjoy\" height=\"32\" src=\"mxc://matrix.example.org/abc\" data-blurhash=\"LEHV6nWB2yk8pyo\" />

Reviewer 3: what about `data-mx-blurhash` to stay in line with all the previous things? (`data-mx-spoiler`, `data-mx-color`, `data-mx-bg-color`)

thumbnails are calculated by media repositories today.

The
[`/_matrix/media/r0/upload`](https://matrix.org/docs/spec/client_server/r0.6.0#post-matrix-media-r0-upload)
endpoint response is modified to include an optional `blurhash` key,
which the client may use to insert into messages if desired:

Example response:

```
{
  "content_uri": "mxc://example.com/abcde123",
  "blurhash": "LKO2?U%2Tw=w]~RB"
}
```

## Visualisation

Viewing an image message that is loading:

![blurhashed preview](images/2448-blurhash.png)

Once the image loads:

![the image has loaded](images/2448-loaded-image.png)

For reference, the current state of things in Riot is:

![boo, sad](images/2448-current-state.png)

## Alternatives

We could include a base64 thumbnail of the image in the event, but blurhash
produces much more efficient textual representations.

## Backwards compatibility

Older clients would ignore the new `blurhash` parameter.

Newer clients would only show it if it exists.

## Unstable prefixes

Implementations wishing to add this before this MSC is merged can do so with
the following:

The `blurhash` key in `m.room.message` should be replaced with
`xyz.amorgan.blurhash`.

> The `blurhash` key in `/_matrix/media/r0/upload`s response should be replaced

Reviewer 5: we also need a `/_matrix/media/unstable/xyz.amorgan/upload` or similar

Author: done

`data-mx-blurhash`, where the value of the attribute is the blurhash
representation of the inline image.

This would be optionally displayed with the inline image itself is loaded in
parallel.

Example `m.room.message.formatted_body`:

```
"formatted_body": This is awesome <img alt=\"flutterjoy\" title=\"flutterjoy\" height=\"32\" src=\"mxc://matrix.example.org/abc\" data-blurhash=\"LEHV6nWB2yk8pyo\" />
```

Note that that the blurhash MUST be HTML-encoded here to comply with HTML
standards.

To be clear: This does not replace thumbnails - it will be shown before they
are downloaded.

## Calculating a blurhash

BlurHashes are inserted into `m.room.message` events by the client, however
some clients may not be able to implement the BlurHash library for whatever
reason. In this case, it would be nice to allow the media repository to
calculate the BlurHash of a piece of media for the client, similar to how
thumbnails are calculated by media repositories today.

The
[`/_matrix/media/r0/upload`](https://matrix.org/docs/spec/client_server/r0.6.0#post-matrix-media-r0-upload)
endpoint response is modified to include an optional `blurhash` key,

> which the client may use to insert into messages if desired:

Reviewer 3: Perhaps also introduce a new media endpoint to return a blurhash? This could be useful for bridges: A bridge sends an inline image, only has the mxc URI available for some reason and wants to get the blurhash from it. Thus it pings the media repo what the blurhash is, instead of having to download the entire file and compute it.

Author: Sounds useful, yes. Any suggestions for API shapes?

I'm thinking:

```
GET /_matrix/media/r0/download/{serverName}/{mediaId}/blurhash
```

Reviewer 3: Yes, and as reply `{"blurhash": "blah"}`

Reviewer 3: Note that such an API call would also generate a blurhash off of non-images etc.

Author: Done.

While thumbnails exist to combat this to some degree, they still need to be
downloaded from a homeserver, which is not instantaneous.

Instead, a BlurHash can be sent inside the `m.room.message` event, which upon
receipt other clients can render for a pretty preview while the actual
thumbnail downloads. They also do not contain any `"` characters, making them
simple to stick inside existing JSON blobs.

## Proposal

### m.room.message

A new optional field is added in `m.room.message`'s content field called
`blurhash`. It is a BlurHash of the original piece of media. Clients could
then render this using [one of the available BlurHash
implementations](https://github.com/woltapp/blurhash).

This would be optionally displayed while the thumbnail of the media is loaded
in parallel.

Example `m.room.message` content field:

```
{
  "body": "image.png",
  "info": {
    "size": 149234,
    "mimetype": "image/png",
    "thumbnail_info": {
      "w": 301,
      "h": 193,
      "mimetype": "image/png",
      "size": 72958
    },
    "w": 301,
    "h": 193,
    "thumbnail_url": "mxc://example.org/abcdefg"
  },
  "msgtype": "m.image",
  "url": "mxc://example.org/abcde",

>   "blurhash": "JadR*.7kCMdnj"

Reviewer 6: Wouldn't this property make more sense, if it was included under the `info` object, since it is basically an alternative to `thumbnail_url` and you need the `w` and `h` anyway to properly preview it.

```
"formatted_body": This is awesome <img alt=\"flutterjoy\" title=\"flutterjoy\" height=\"32\" src=\"mxc://matrix.example.org/abc\" data-blurhash=\"LEHV6nWB2yk8pyo\" />
```

To be clear: This does not replace thumbnails - it will be shown before they
are downloaded.

## Calculating a blurhash

BlurHashes are inserted into `m.room.message` events by the client, however
some clients may not be able to implement the BlurHash library for whatever
reason. In this case, it would be nice to allow the media repository to
calculate the BlurHash of a piece of media for the client, similar to how
thumbnails are calculated by media repositories today.

The
[`/_matrix/media/r0/upload`](https://matrix.org/docs/spec/client_server/r0.6.0#post-matrix-media-r0-upload)
endpoint response is modified to include an optional `blurhash` key,
which the client may use to insert into messages if desired:

Example response:

```
{
  "content_uri": "mxc://example.com/abcde123",
  "blurhash": "LKO2?U%2Tw=w]~RB"
}
```

In addition, a new endpoint will be added to the Media API that allows a
client with an existing mxc URL to retrieve a blurhash for it from the
server. It will take the form of `GET

> /_matrix/media/r0/download/{serverName}/{mediaId}/blurhash`.

Reviewer 5: please not download, as `blurhash` is a valid file name. Ideally this would show up in https://github.com/matrix-org/matrix-doc/pull/2380 not here.

Author: Ah, yes, poor shape choice.

Would you be happy with a temporary API for this in the meantime until #2380 lands?

Author: How about `GET /_matrix/media/r0/blurhash/{serverName}/{mediaId}` instead?

Reviewer 5: I'd rather we pushed 2380 through tbh

All this MSC needs to do is say "also MSC2380, or similar, will have a `blurhash` property`. A dedicated endpoint for such a value is not great.

Author: Huh, yeah that could be useful, though if a client doesn't support it then is having the server render it and then send it down any different from pulling down a thumbnail?

Reviewer 5: In theory the blurhash would be faster to calculate and return, much like how identicons have a fairly linear transfer time.

Author: Ah I see, you mean send the blurhash string rather than the computed PNG. Yeah that'd be nice.

Author: Right, I misread your comment. So you're suggesting hold off on having the server provide a dedicated blurhash endpoint in favour of MSC2380. That sounds fair, and will just mean clients will need to compute it in implementation for now.

Is that correct?

Author: Ah, one issue with using MSC2380, and not a dedicated endpoint, is that clients won't be able to specify the X/Y parameters of the resolution of the blurhash (or you could, but adding blurhash-specific parameters to `/info` seems a bit weird).

Reviewer 6: I don't think not being able to specify the blurhash components is a major issue. You need to know the image anyway, to specify those parameters, so it is probably best if the media repo does that.

Reviewer 5: Yea the thing generating the blurhash would know the components and the components can be pulled out of the blurhash. I don't think we really need a way to specify the components as that actually adds overhead (the media repo then has to decide if it cares, and possibly recalculate it).

MSC2380 is just to return a bunch of stuff about an uploaded file, such as the blurhash. A separate download endpoint (which takes a blurhash string) could render whatever it was given faster than trying to deal with thumbnailing. Separately to both of those, the upload endpoint could maybe take X and Y components, but I'd personally be inclined to just ignore those in my implementation because uploading something is already hard enough on the server.

Author: Ok, so no component specification.

I've updated the doc. Let me know if we're on the same page?

> /_matrix/media/r0/download/{serverName}/{mediaId}/blurhash`.

Reviewer 5: A different option might be `/_matrix/media/r0/blurhash/{encodedBlurhash}?width=512&height=512` which forces the server to render it instead.

Author: I kind of like that idea, though it could be a potential exploit vector if there was a vuln in the blurhash library implementation, and maybe a DOS vector if you give it a really big blurhash string.

Author: Huh, ok, https://github.com/matrix-org/matrix-doc/pull/2448#discussion_r386068301 should've ended up in this thread. Repeating it here instead:

> Huh, yeah that could be useful, though if a client doesn't support it then is having the server render it and then send it down any different from pulling down a thumbnail?

Basically I'm happy to have an endpoint for returning the blurhash string from an mxc URL, though not sure about an image from a blurhash string, due to potential attack vectors and that downloading a blurhash as an image would probably not be much faster than downloading the thumbnail itself.

Author: We've decided to go for the extra endpoint. See https://github.com/matrix-org/matrix-doc/pull/2448#discussion_r385849454.

Example request:

```
GET /_matrix/media/r0/download/example.com/abcde12345/blurhash
```

Example response:

```
{
  "blurhash": "LEHV6nWB2yk8pyo"
}
```

## Visualisation

Viewing an image message that is loading:

![blurhashed preview](images/2448-blurhash.png)

Once the image loads:

![the image has loaded](images/2448-loaded-image.png)

For reference, the current state of things in Riot is:

![boo, sad](images/2448-current-state.png)

## Alternatives

We could include a base64 thumbnail of the image in the event, but blurhash
produces much more efficient textual representations.

## Backwards compatibility

Older clients would ignore the new `blurhash` parameter.

Newer clients would only show it if it exists.

> ## Unstable prefixes

Reviewer 5: also an unstable feature flag to detect the endpoint, which I assume would be `xyz.amorgan.blurhash` in this convention?

Author: Can one not just see if the endpoints `404`, or do we want to add the flag to save a network request?

Reviewer 5: Detection over 404 just means more round trips and sketchy code. We've historically not wanted to use 404 as a detection means and invented flags for this purpose.

Author: Makes sense. Added.

downloaded from a homeserver, which is not instantaneous.

Instead, a BlurHash can be sent inside the `m.room.message` event, which upon
receipt other clients can render for a pretty preview while the actual
thumbnail downloads. They also do not contain any `"` characters, making them
simple to stick inside existing JSON blobs.

To be clear: A BlurHash does not replace a thumbnail - it will be shown
before the thumbnail is downloaded.

## Proposal

### m.room.message

A new optional field is added in `m.room.message`'s content field called
`blurhash`. It is a BlurHash of the original piece of media. Clients could
then render this using [one of the available BlurHash
implementations](https://github.com/woltapp/blurhash).

This would be optionally displayed while the thumbnail of the media is loaded
in parallel.

Example `m.room.message` content field:

```
{
  "body": "image.png",
  "info": {
    "size": 149234,
    "mimetype": "image/png",
    "thumbnail_info": {
      "w": 301,
      "h": 193,
      "mimetype": "image/png",
      "size": 72958
    },
    "w": 301,
    "h": 193,
    "thumbnail_url": "mxc://example.org/abcdefg",

>     "blurhash": "JadR*.7kCMdnj"

Reviewer 5: should it be encrypted?

Reviewer 7: If you mean should it be present in encrypted image messages, I'd say yes, there's no reason not to. You obviously can't use the media repo to generate it for encrypted images, but that's how normal thumbnails work too. The libraries used to render blurhashes client-side can usually generate them too.

Reviewer 6: I think this should be encrypted, when the other properties of this (apart from thumbnail_url) are too, like h and w.

Reviewer 3: Why do we need to special-ase the `blurhash` property for that? The entire event content gets encrypted with megolm after all, right?

Reviewer 5: Encrypted files have a different format.

Reviewer 3: Encrypted files also have an `info` block, just as unencrypted files have, so you just toss the blurhash in there, as this MSC specifies. not sure why you have to explicitly say that it lands in `info` for both encrypted and unencrypted files, as the `info` format is the same?

Reviewer 5: from a spec perspective they're different. Common sense says they should be the same, but all the MSC needs to say is something like "because the `blurhash` property is under `info`, it can be encrypted to avoid leaking data"

Reviewer 3: [the spec](https://matrix.org/docs/spec/client_server/latest#m-image) does not do any differentiation between encrypted and unencrypted info blocks at all. It seems unneeded to add it here?

Reviewer 3: If at all, this MSC could mention under "security considerations" something like "as the blurhash is in the `info` block of the event content and the entire event content gets encrypted in e2ee room, no metadata is leaked in those"

Reviewer 6: > Common sense says they should be the same, but all the MSC needs to say is something like "because the blurhash property is under info, it can be encrypted to avoid leaking data"

That sounds a bit unnecessary. Everything in the event needs to be encrypted. There are only a few specific things that aren't encrypted, specifically `relates_to`, which is explicitly called out in the spec. I hope every client encrypts the whole content and only adds relations to the unencrypted content, which is still sent. You need a specific allow list for what NOT to encrypt, not the other way around! I don't see any reason to assume something would not be encrypted, unless explicitly stated so.

Reviewer 4: I agree with @deepbluev7: the default for encrypted messages is that whatever is in `contents` gets encrypted.  Anything that should not be encrypted (or that should be both in the encrypted and non-encrypted parts) should be listed as exceptions, but for everything else, it should be understood as being encrypted.  If we need to add words to the spec to make that clearer, then we can do so.

Author: I agree that it should not be required to mention it, though I see no harm in doing so to add clarification. Done in 3994010.

  },
  "msgtype": "m.image",
  "url": "mxc://example.org/abcde",
}
```

Note that a BlurHash representation is really only applicable to media, and
as such should only be used in conjunction with the following
`m.room.message` msgtypes:

* `m.image`
* `m.video`

### Inline images

An optional attribute is added to `<img>` tags in messages:
`data-mx-blurhash`, where the value of the attribute is the blurhash
representation of the inline image.

This would be optionally displayed while the inline image itself is loaded in
parallel.

Example `m.room.message.formatted_body`:

```

> "formatted_body": This is awesome <img alt=\"flutterjoy\" title=\"flutterjoy\" height=\"32\" src=\"mxc://matrix.example.org/abc\" data-blurhash=\"LEHV6nWB2yk8pyo\" />

Reviewer 3: ```suggestion
"formatted_body": This is awesome <img alt=\"flutterjoy\" title=\"flutterjoy\" height=\"32\" src=\"mxc://matrix.example.org/abc\" data-mx-blurhash=\"LEHV6nWB2yk8pyo\" />
```

```
<bytes representing BlurHash image>
```

In addition, the server can return the BlurHash string for an image when
given an MXC URL. This would be through something like the Media Information
API (specified in
[MSC2380](https://github.com/matrix-org/matrix-doc/pull/2380)), or similar.

## Visualisation

Viewing an image message that is loading:

![blurhashed preview](images/2448-blurhash.png)

Once the image loads:

![the image has loaded](images/2448-loaded-image.png)

For reference, the current state of things in Riot is:

![boo, sad](images/2448-current-state.png)

## Alternatives

We could include a base64 thumbnail of the image in the event, but blurhash
produces much more efficient textual representations.

## Backwards compatibility

Older clients would ignore the new `blurhash` parameter.

Newer clients would only show it if it exists.

## Unstable prefixes

Implementations wishing to add this before this MSC is merged can do so with
the following:

The `blurhash` key in `m.room.message` should be replaced with
`xyz.amorgan.blurhash`.

`/_matrix/media/r0/upload` should be replaced with
`/_matrix/media/unstable/xyz.amorgan/upload`, which keeps the same `blurhash`

> response key.

Reviewer 5: I'd more recommend that the `/upload` endpoint return a `xyz.amorgan.blurhash` property instead of defining a new endpoint.

Reviewer 5: ftr this is what the web implementation relies on: https://github.com/matrix-org/matrix-react-sdk/pull/5099

Author: Yes, that would be cleaner thinking about it.

Author: Done in 616bc81.

This would be optionally displayed while the inline image itself is loaded in
parallel.

Example `m.room.message.formatted_body`:

```
"formatted_body": This is awesome <img alt=\"flutterjoy\" title=\"flutterjoy\" height=\"32\" src=\"mxc://matrix.example.org/abc\" data-blurhash=\"LEHV6nWB2yk8pyo\" />
```

## Calculating a blurhash

BlurHashes are inserted into `m.room.message` events by the client, however
some clients may not be able to implement the BlurHash library for whatever
reason. In this case, it would be nice to allow the media repository to
calculate the BlurHash of a piece of media for the client, similar to how
thumbnails are calculated by media repositories today.

The
[`/_matrix/media/r0/upload`](https://matrix.org/docs/spec/client_server/r0.6.0#post-matrix-media-r0-upload)
endpoint response is modified to include an optional `blurhash` key,
which the client may use to insert into messages if desired:

Example response:

```
{
  "content_uri": "mxc://example.com/abcde123",
  "blurhash": "LKO2?U%2Tw=w]~RB"
}
```

A server MUST also be able to render a BlurHash image from a BlurHash string.
This is to support clients that do not have a BlurHash client implementation

> or are otherwise unable to do so themselves.

Reviewer: I am unconvinced that this is a use case worth supporting, if not least because it seems counter to the goal of blurhashing itself. You might be able to make an argument as an optional extra, but it certainly shouldn't be a MUST.

First and foremost, if I correctly understood the point of a blurhash, it's that it should be a filler while the client waits for a thumbnail to be loaded. The blurhash needs to be nearly instantaneous in order to be useful, because any amount of appreciable time spent in the "empty box" state is going to be a detractor from the experience of a client.
 
Consider that it might take 50ms for the mediarepo to render a blurhash and return it to the client, and concurrently take 100ms to return a thumbnail. You'll end up with 50ms of whitespace, 50ms of blurhash, and then the thumbnail. It will look more like a flicker effect than an actual placeholder!

Therefore, it makes little sense for the server to process a blurhash. You'll increase the number of requests needed to be made per image (think of the avatars!), increase the expense of resources on the server, and you'll end up with more delay. Even if blurhashes take a few ms to process, a server like matrix.org or mozilla.org will be queueing requests and won't be completing them even close to the mark.

Client libraries for major languages exist in abundance (and even officially supported) at https://github.com/woltapp/blurhash. Even if your language does not support it, the algorithm is [fairly simple to implement](https://github.com/woltapp/blurhash/blob/master/Algorithm.md).

I would argue that it's perfectly fine for clients to *not* render a blurhash if they are unable to do so, rather than relying on the server to do it for them. The implementation complexity of a new endpoint, and the resource requirements that come with that, are simply not worth it.

Reviewer 5: I'd agree that this doesn't need to be a MUST, but I would still recommend it being in the MSC. Thumbnails do not take 100ms to render, particularly for 2-4mb images that people love posting. Unless you're lucky enough that the media is in memory and the thumbnail has already been generated, of course. But if you're not in the 2% of requests and having to go over federation your average request time is in the seconds range, not milliseconds. 

Having the endpoint is more realistically useful for simpler client implementations. If you don't want to add a whole library to your client for one feature because you're concerned about bloat you can settle for a spinner + blurhash request + thumbnail request. I would expect that Riot and other popular clients would do it all client side though instead of asking the server to do it.

Author: You make a fair point. @turt2live was there any other reason to have the server render these other than for clients that couldn't do so?

Reviewer: I'm okay with it being optional (although, that causes fun and games on how to determine capabilities for a media server). I would like to see how the PoCs do in terms of how they feel to an end user.

Reviewer 5: The only other reason is so media repos can validate their implementations without having to tie themselves to a client. Having it be optional (or required, but unused) is fine by me, largely for the concerns I've outlined already a number of times.

Reviewer 6: As a client developer, I probably wouldn't want to use the optional blurhash download route. It adds complexity for really not that much benefit. Also blurhash rendering is not that fast either, so I don't think such an endpoint would be much faster than requesting the thumbnail directly and the requests could even end up fighting for resources on client, network and server. Is the extra complexity of an optional endpoint really worth the gains in this case?

Author: > The only other reason is so media repos can validate their implementations without having to tie themselves to a client.

@turt2live I don't think I'm grokking what you mean here. Do you mean their implementations of the blurhash algorithm? If so, can't that be validated against test vectors already provided?

Reviewer 5: sorry, forgot about this thread ._. 

It could be validated, yes. It could also be validated by the implementation adding the endpoint itself.

fwiw I'm more aligned with @Half-Shot these days after time has passed: this doesn't seem to be needed in the spec.

Reviewer 8: I'm with @Half-Shot here - I love the idea of blurhash (first time I've heard of it tbh) and it's trivially easy to add a `blurhash` key to `/upload` on Dendrite. Adding a blurhash download endpoint is only useful if the remote file is not on the local server and the client is not capable of rendering blurhashes. Speed is important here, and the bulk of the download on a HS for a thumbnail will likely be RTT and not data transfer - given both Synapse and Dendrite pre-calculate thumbnails on upload so there's no extra computation to the download request.

I also feel it adds unnecessary complexity to rendering images and may even slow down image rendering if clients don't do requests in parallel e.g:
 - Get message JSON
 - Read w/h and set image bounds on UI
 - Make request to blurhash
 - Receive blurhash response
 - Render blurhash image
 - Make request to thumbnail
 - Receive thumbnail response
 - Render thumbnail response

If the bulk of the time is on RTT then this will be slowly than just going straight to the thumbnail, unless you do things in parallel.

Author: Sounds like we've reached a conclusion, which I agree with. I've removed the endpoint. Media repositories/homeserver will no longer need to implement `/_matrix/media/r0/blurhash/{hash}`.

```

In addition, the server can return the BlurHash string for an image when
given an MXC URL. This would be through something like the Media Information
API (specified in
[MSC2380](https://github.com/matrix-org/matrix-doc/pull/2380)), or similar.

## Visualisation

Viewing an image message that is loading:

![blurhashed preview](images/2448-blurhash.png)

Once the image loads:

![the image has loaded](images/2448-loaded-image.png)

For reference, the current state of things in Riot is:

![boo, sad](images/2448-current-state.png)

## Alternatives

We could include a base64 thumbnail of the image in the event, but blurhash
produces much more efficient textual representations.

## Backwards compatibility

Older clients would ignore the new `blurhash` parameter.

Newer clients would only show it if it exists.

## Unstable prefixes

Implementations wishing to add this before this MSC is merged can do so with
the following:

The `blurhash` key in `m.room.message` should be replaced with
`xyz.amorgan.blurhash`.

`/_matrix/media/r0/upload` should be replaced with
`/_matrix/media/unstable/xyz.amorgan/upload`, which keeps the same `blurhash`
response key.

> `/_matrix/media/r0/blurhash/{serverName}/{mediaId}` should be

Reviewer 5: ```suggestion
`/_matrix/media/r0/blurhash/{hash}` should be
```

Author: This endpoint has been removed.

replaced with

> `/_matrix/media/unstable/xyz.amorgan/blurhash/{serverName}/{mediaId}`.

Reviewer 5: ```suggestion
`/_matrix/media/unstable/xyz.amorgan/blurhash/{hash}`.
```

Author: Well, this endpoint is for getting a blurhash string from a mxc url, so why `{hash}`?

Reviewer 5: because the MSC makes no mention of this endpoint, but does make mention of the hash endpoint...

Author: This endpoint has been removed.

> # MSC2448: Using BlurHash in Media Events

Reviewer 3: How about adding blurhashes to URL previews, too?

Author: Oh, that's not a bad idea. I'm hesitant to scope creep this too much as we've already got multiple implementations floating around, but it shouldn't be a massive implementation change on either the server or client side... so why not.

Author: Added in e0a74423.

[BlurHash](https://blurha.sh) is a compact representation of a placeholder
for an image (or the frame of a video). Currently in Matrix, clients must
display a placeholder image in the message timeline while a piece of media is
loading. Some clients, such as Riot, simply display an empty space.

While thumbnails exist to combat this to some degree, they still need to be
downloaded from a homeserver, which is not instantaneous.

Instead, a BlurHash can be sent inside the `m.room.message` event, which upon
receipt other clients can render for a pretty preview while the actual
thumbnail downloads. They also do not contain any `"` characters, making them
simple to stick inside existing JSON blobs.

To be clear: A BlurHash does not replace a thumbnail - it will be shown
before the thumbnail is downloaded.

## Proposal

### m.room.message

> A new optional field is added in `m.room.message`'s content field called

Reviewer 9: Whilst implementing this for Ruma, I noticed we have a matching ImageInfo type in `m.sticker` and `m.room.avatar`. Should the `blurhash` field be added to those too?

Reviewer: Feels like they would be relevant there, especially stickers.

Author: Both sound useful.

Author: After talking with @Half-Shot he mentioned that we already add extra fields to `m.room.member`, such as [is_direct](https://matrix.org/docs/spec/client_server/r0.6.1#id143) in invite member events.

So adding another optional field for blurhash is probably fine.

Author: Done in https://github.com/matrix-org/matrix-doc/pull/2448/commits/b761b069ab060c334319df60f841346db03edb25, https://github.com/matrix-org/matrix-doc/pull/2448/commits/40d71ff4a229826da3b3621f1c0dad35a82fbd43 and https://github.com/matrix-org/matrix-doc/pull/2448/commits/7f131843383fc0c1126bb6ac4320d3558b2d6b72.

}
```

Note that a BlurHash representation is really only applicable to media, and
as such should only be used in conjunction with the following
`m.room.message` msgtypes:

* `m.image`
* `m.video`

### Inline images

An optional attribute is added to `<img>` tags in messages:
`data-mx-blurhash`, where the value of the attribute is the blurhash
representation of the inline image.

This would be optionally displayed while the inline image itself is loaded in
parallel.

Example `m.room.message.formatted_body`:

```
"formatted_body": This is awesome <img alt=\"flutterjoy\" title=\"flutterjoy\" height=\"32\" src=\"mxc://matrix.example.org/abc\" data-mx-blurhash=\"LEHV6nWB2yk8pyo\" />
```

## Calculating a blurhash

BlurHashes are inserted into `m.room.message` events by the client, however
some clients may not be able to implement the BlurHash library for whatever
reason. In this case, it would be nice to allow the media repository to
calculate the BlurHash of a piece of media for the client, similar to how
thumbnails are calculated by media repositories today.

The
[`/_matrix/media/r0/upload`](https://matrix.org/docs/spec/client_server/r0.6.0#post-matrix-media-r0-upload)

> endpoint response is modified to include an optional `blurhash` key,

Reviewer 6: I really like this feature and I would love to see it move forward. I'm not really sure, what's currently blocking that though. To me it seems like the client side features are pretty clear and also widely used by clients. On the other hand, while the media repo implements the blurhash endpoints, that hasn't seen much use in clients. Additionally it seems like most discussion happens around the server endpoints, their usefulness and their intended use.

So I would suggest dropping the server side features from this MSC. If they are actually useful, they can be added at a later point, but without them this MSC is really small and achieves 80% of its intended purpose. It should be fairly simple for the SCT to review then.

I know that is a bold suggestion and I appreciate the thought, that went into the server endpoints, but I would really love it, if all clients would use this feature sooner rather than later and I think cutting the more controversial features does not diminish the use of this MSC much.

Does that sound reasonable?

Author: The controversial feature of the media repo being able to render an image from a BlurHash string has been dropped. However the optional requirement for the media repo to return a BlurHash string after a successful `/_matrix/media/r0/upload` call is still present as it seemed generally agreeable to everyone.

produces much more efficient textual representations.

## Backwards compatibility

Older clients would ignore the new `blurhash` parameter.

Newer clients would only show it if it exists.

## Unstable prefixes

Implementations wishing to add this before this MSC is merged can do so with
the following:

The `blurhash` key in `m.room.message` should be replaced with
`xyz.amorgan.blurhash`.

`/_matrix/media/r0/upload` should return a `xyz.amorgan.blurhash` key
containing the blurhash instead of `blurhash`. This is preferred to adding
another endpoint, which is messier to later remove.

`/_matrix/media/r0/blurhash/{serverName}/{mediaId}` should be
replaced with
`/_matrix/media/unstable/xyz.amorgan/blurhash/{serverName}/{mediaId}`.

The `data-mx-blurhash` attribute in `<img>` tags should be replaced with
`data-xyz-amorgan-blurhash`.

And finally, an entry should be added to the homeserver's `GET
/_matrix/client/versions` endpoint, in `unstable_features`, with the key
`xyz.amorgan.blurhash` set to `true`.

## Security considerations

Blurhash entries in encrypted events, be it as part of the `info` property,
or `<img>` tags, should be encrypted along with the rest of the event

> content.

Author: The endpoint allowing clients to outsource rendering of a blurhash to a media server has been removed, eliminating this concern :slightly_smiling_face:

      "w": 301,
      "h": 193,
      "mimetype": "image/png",
      "size": 72958
    },
    "w": 301,
    "h": 193,
    "thumbnail_url": "mxc://example.org/abcdefg",
    "blurhash": "JadR*.7kCMdnj"
  },
  "msgtype": "m.image",
  "url": "mxc://example.org/abcde",
}
```

Note that a BlurHash representation is really only applicable to media, and
as such should only be used in conjunction with the following
`m.room.message` msgtypes:

* `m.image`
* `m.video`

### m.sticker

An optional field is added to `m.sticker`'s `content.info` dictionary with
the key `blurhash`. Its value is a BlurHash of the sticker media.

Example `m.sticker` content:

```json
{
  "body": "Landing",
  "info": {
    "h": 200,
    "mimetype": "image/png",
    "size": 73602,
    "thumbnail_info": {
      "h": 200,
      "mimetype": "image/png",
      "size": 73602,
      "w": 140
    },
    "thumbnail_url": "mxc://matrix.org/sHhqkFCvSkFwtmvtETOtKnLP",
    "w": 140,
    "blurhash": "JadR*.7kCMdnj"
  },
  "url": "mxc://matrix.org/sHhqkFCvSkFwtmvtETOtKnLP"
}
```

### m.room.avatar

Room avatars having BlurHashes available will be especially useful when
viewing a server's Public Rooms directory.

> An optional field is added to `m.room.avatar`'s `content` dictionary with the

Reviewer 10: `m.room.avatar` has an optional `info` object with all-optional fields like the aforementioned events too. Why would the blurhash be part of the `content` object rather than `content.info` here?

Author: So it does! That was an oversight then, thanks for pointing it out!

# MSC2448: Using BlurHash as a Placeholder for Matrix Media

[BlurHash](https://blurha.sh) is a compact representation of a placeholder

> for an image (or the frame of a video). Currently in Matrix, clients must

Reviewer 11: ```suggestion
for an image (or a frame of video). Currently in Matrix, clients must
```

```json
{
  "url": "mxc://amorgan.xyz/a59ee02f180677d83d1b57d366127f8e1afdd4ed",
  "info": {
    "blurhash": "JadR*.7kCMdnj"
  }
}
```

### m.room.member

Much like room avatars, user avatars can have BlurHashes as well. There is a
little more required to implement this, but the outcome of no longer having
missing avatars upon opening a room is worthwhile.

An optional field is added to `m.room.member`'s `content` dictionary with
the key `blurhash`. Its value is a BlurHash of the media that is pointed
to by `avatar_url`.

Note that `blurhash` SHOULD be omitted if `avatar_url` is not present.

Example `m.room.member` event content:

```json
{
  "avatar_url": "mxc://example.org/SEsfnsuifSDFSSEF",
  "displayname": "Alice Margatroid",
  "membership": "join",
  "blurhash": "JadR*.7kCMdnj"
}
```

### Profile endpoints

Endpoints that return profile information, and thus MXC URLs to user avatars, are
extended to optionally include BlurHashes as well.

[`GET /_matrix/client/r0/profile/{userId}`](https://matrix.org/docs/spec/client_server/r0.6.1#get-matrix-client-r0-profile-userid) has an optional field added with
the key `blurhash`. Its value is a BlurHash of the media that is pointed to

> by `avatar_url`. `blurhash` MUST be omitted if `avatar_url` is not present.

Author: Good spot! That is indeed an inconsistency. Other than the fact that we already have a majority of MUST, I believe we should use MUST instead of SHOULD as an event that only has a blurhash and not an image would be a confusing user experience (assuming the client did not detect the lack of image).

A user may see a blurhash and be waiting for the corresponding image to load. Yet if there is no corresponding image in the event, the user may assume their media repository is faulty, rather than the event being malformed. Putting MUST here explicitly states such an event is invalid.

> by `avatar_url`. `blurhash` MUST be omitted if `avatar_url` is not present.

Author: Fixed in 9bd0ba631f48b46feca7b0d2fce0e8ba055c70d3.

The same applies to [`GET /_matrix/client/r0/profile/{userId}/avatar_url`](https://matrix.org/docs/spec/client_server/r0.6.1#get-matrix-client-r0-profile-userid-avatar-url), and to the federation endpoint [`GET /_matrix/federation/v1/query/profile`](https://matrix.org/docs/spec/server_server/r0.1.4#get-matrix-federation-v1-query-profile).

[`PUT /_matrix/client/r0/profile/{userId}/avatar_url`](https://matrix.org/docs/spec/client_server/r0.6.1#put-matrix-client-r0-profile-userid-avatar-url) has an optional field added
to the request body with the key `blurhash`. Its value is a BlurHash of the media that is pointed to by the value of the `avatar_url` field in the same request.

### URL previews

An optional attribute is added to the OpenGraph data returned by a call
to
[`GET /_matrix/media/r0/preview_url`](https://matrix.org/docs/spec/client_server/r0.6.1#get-matrix-media-r0-preview-url)

> called `matrix:image:blurhash`. The value

Author: I put it under the `matrix` namespace as the [OpenGraph specification](https://ogp.me/) doesn't include an `og:image:blurhash` entry (yet!).

I've added a clarification for this in 75a4fa6c.

  "og:image:type": "image/png",
  "og:image:height": 48,
  "og:image:width": 48,
  "matrix:image:size": 102400,
  "matrix:image:blurhash": "oyp8ky2BWn7VHEL"
}
```

### Inline images

An optional attribute is added to `<img>` tags in messages:
`data-mx-blurhash`, where the value of the attribute is the blurhash
representation of the inline image.

This would be optionally displayed while the inline image itself is loaded in
parallel.

Example `m.room.message.formatted_body`:

```
"formatted_body": This is awesome <img alt=\"flutterjoy\" title=\"flutterjoy\" height=\"32\" src=\"mxc://matrix.example.org/abc\" data-mx-blurhash=\"LEHV6nWB2yk8pyo\" />
```

## Calculating a blurhash

BlurHashes are inserted into `m.room.message` events by the client, however
some clients may not be able to implement the BlurHash library for whatever
reason. In this case, it would be nice to allow the media repository to
calculate the BlurHash of a piece of media for the client, similar to how
thumbnails are calculated by media repositories today.

The
[`/_matrix/media/r0/upload`](https://matrix.org/docs/spec/client_server/r0.6.0#post-matrix-media-r0-upload)

> endpoint response is modified to include an optional `blurhash` key,

Reviewer 12: I would like a way for a client to be able to signal for the media repo to not bother generating a blurhash, Element clients will likely always be generating them locally, as well as their thumbnails, no point wasting CPU cycles and progressing global warming

Reviewer 6: Nheko doesn't use that endpoint either. In fact, I don't know any client that does, probably only terminal/curl clients will?

Reviewer 12: Heck, maybe it should even be opt-in, given any non-supporting clients won't be using the field in the first place

Author: Good point, yes. I expect the majority of clients, once they support blurhashes, will likely be generating them themselves.

If a client does not support blurhashes then it won't be doing anything with this new `blurhash` response key anyways. So this really should be opt-in then.

Author: Done in 9e981ba7.

Users who have not specified `blurhash` in their `m.room.member` event yet may
stand out from users who have while both are loading. This is entirely up to
clients to handle, though a suggestion may be to "fake" a blurhash by
blurring some placeholder image (perhaps something with variation between
users like Element's coloured backgrounds with letters in them, or [an
identicon](https://en.wikipedia.org/wiki/Identicon) derived from the user's
ID) until the user's actual avatar loads.

## Unstable prefixes

Implementations wishing to add this before this MSC is merged can do so with
the following:

* The `blurhash` key in any events, request or response bodies should be
replaced with `xyz.amorgan.blurhash`.

* `/_matrix/media/r0/upload` should return a `xyz.amorgan.blurhash` key
containing the blurhash instead of `blurhash`. This is preferred to adding
another endpoint, which is messier to later remove.

* The `data-mx-blurhash` attribute in `<img>` tags should be replaced with
`data-xyz-amorgan-blurhash`.

* And finally, an entry should be added to the homeserver's `GET
/_matrix/client/versions` endpoint, in `unstable_features`, with the key

> `xyz.amorgan.blurhash` set to `true`.

Author: Oops, it was required (well not really, but it was at least somewhat useful) when we had a new endpoint for calculating blurhashes in an earlier version of this document. But since that's now been removed, so can this requirement.

# MSC2448: Using BlurHash as a Placeholder for Matrix Media

[BlurHash](https://blurha.sh) is a compact representation of a placeholder
for an image (or a frame of video). Currently in Matrix, clients must
display a placeholder image in the message timeline while a piece of media is
loading. Some clients, such as Element, simply display an empty space.

While thumbnails exist to combat this to some degree, they still need to be
downloaded from a homeserver, which is not instantaneous.

> Instead, a BlurHash can be sent inside the `m.room.message` event, which upon

Reviewer 5: because stickers:
```suggestion
Instead, a BlurHash can be sent inside message-like events, which upon
```

receipt other clients can render for a pretty preview while the actual
thumbnail downloads. They also do not contain any `"` characters, making them
simple to stick inside existing JSON blobs.

To be clear: A BlurHash does not replace a thumbnail - it will be shown

> before the thumbnail is downloaded.

Reviewer 13: It'd be good to spell out how blurhashes should interact with progressively loaded thumbnails, and make UX recommendations on how to spell out that they are placeholders rather than the actual thumbnail - i.e. the finessing we just did on Eleweb, to remind implementors they'll need to do the same.

## Proposal

### m.room.message

An optional field is added in `m.room.message`'s `content.info` dictionary
with the key `blurhash`. It is a BlurHash of the original piece of media.
Clients could then render this using [one of the available BlurHash
implementations](https://github.com/woltapp/blurhash).

This would be optionally displayed while the thumbnail of the media is loaded
in parallel.

Example `m.room.message` content:

```json
{
  "body": "image.png",
  "info": {
    "size": 149234,
    "mimetype": "image/png",
    "thumbnail_info": {
      "w": 301,
      "h": 193,
      "mimetype": "image/png",
      "size": 72958
    },
    "w": 301,
    "h": 193,
    "thumbnail_url": "mxc://example.org/abcdefg",
    "blurhash": "JadR*.7kCMdnj"
  },
  "msgtype": "m.image",
  "url": "mxc://example.org/abcde",
}
```

Note that a BlurHash representation is really only applicable to media, and
as such should only be used in conjunction with the following
`m.room.message` msgtypes:

* `m.image`
* `m.video`

> ### m.sticker

Reviewer 12: Slight hurdle, blurhash doesn't support transparencies so doesn't seem appropriate for Stickers

![image](https://user-images.githubusercontent.com/2403652/119324994-569a9200-bc78-11eb-998e-f632fdec2ae5.png)
![image](https://user-images.githubusercontent.com/2403652/119325010-5a2e1900-bc78-11eb-90fb-9e77b03d1746.png)

The reference site breaks when given a transparency
![image](https://user-images.githubusercontent.com/2403652/119325044-61552700-bc78-11eb-9e4a-67cb702385c0.png)

https://github.com/woltapp/blurhash/issues/100

Reviewer 5: random thought: it might be worth defining a "shape" next to the blurhash where a polygon of max 12 points (or whatever) can be defined

Reviewer 12: Sounds like you want to maintain cross-platform libraries which extend blurhash with generic path masking :D

Reviewer 5: I now realize this is a no-go. A square sticker isn't *that* bad: clients might want to put a circle or something around it as a border to match the rough shape of most stickers.

Reviewer 11: One hacky solution is to generate a blurhash on black and a blurhash on white, the client can then pick the once closest to the theme. But that doubles the storage space. If you wanted to get super fancy a client can probably diff the two and tease the transparency info out. But I don't want to mandate that in the spec. (I would rather wait for a blurhashv2 to come out with transparency support).

Reviewer 14: I actually like @turt2live's proposal, Telegram also stores a rough outline (say 100 points or so), then smooths it out, to make sure the sticker is still recognizable before the "colours" load in.

Reviewer 15: It's my opinion that we can leave adding a "rough outline of a path" as an additional optimization for the future.

Reviewer 5: tbh I disagree, only because stickers look broken with a full-frame blurhash.

Author: The current state of blurhash not supporting transparency (and little movement on this front from the upstream library) is a good sign to potentially back away from supporting blurhashes for `m.sticker` events at the moment, and leave it as a separate optimisation as @sumnerevans mentioned.

The path data sounds like a potential solution, and other messages services such as Telegram indeed use that method to great effect. Some thought will need to be given to animated stickers though, and what frame should be used when capturing a silhouette of the media (the first frame would not always be effective).

Blurhashes in a sticker picker also seems less useful than silhouettes when searching for a known sticker on a slow network connection.

Reviewer 16: I'm curious whether ThumbHash would address this sufficiently: https://github.com/evanw/thumbhash

    "blurhash": "JadR*.7kCMdnj"
  },
  "url": "mxc://matrix.org/sHhqkFCvSkFwtmvtETOtKnLP"
}
```

### m.room.avatar

Room avatars having BlurHashes available will be especially useful when
viewing a server's Public Rooms directory.

An optional field is added to `m.room.avatar`'s `content.info` dictionary with the
key `blurhash`. Its value is a BlurHash of the media that is pointed to by
`url`.

Example `m.room.avatar` content:

```json
{
  "url": "mxc://amorgan.xyz/a59ee02f180677d83d1b57d366127f8e1afdd4ed",
  "info": {
    "blurhash": "JadR*.7kCMdnj"
  }
}
```

### m.room.member

Much like room avatars, user avatars can have BlurHashes as well. There is a
little more required to implement this, but the outcome of no longer having
missing avatars upon opening a room is worthwhile.

An optional field is added to `m.room.member`'s `content` dictionary with
the key `blurhash`. Its value is a BlurHash of the media that is pointed
to by `avatar_url`.

Note that `blurhash` MUST be omitted if `avatar_url` is not present.

Example `m.room.member` event content:

```json
{
  "avatar_url": "mxc://example.org/SEsfnsuifSDFSSEF",
  "displayname": "Alice Margatroid",
  "membership": "join",

>   "blurhash": "JadR*.7kCMdnj"

Reviewer 5: hmm: why is this on the top level but the room avatar isn't? should we just move the room avatar one up a level and call it good?

}

> ```

Reviewer 12: @ara4n raised a point internally that Element Web at least probably wouldn't want to waste CPU cycles on rendering the blurhashes for many tens of avatars on-screen at a time given the server thumbnail is so quick to load and heavily cached

Reviewer 4: Sometimes avatars can take a while to load, and if there are many on-screen, then the client's HTTP connection limit can be reached, making it even slower.  Clients can use some sort of heuristic to decide when to use the blurhash, e.g. if it the avatar has been loading for >1s, then use the blurhash.

Author: It may also be useful when showing an avatar on someone's profile when you click on them from the timeline. Of course it's up to the client to pull that avatar from `/profile` or the membership event.

Author: @t3chguy I'm curious if Element Web still holds the same opinion now that some time has passed.

Reviewer 12: It wasn't EWT's opinion, but that of @ara4n

An example request from a client that would like the server to generate a
blurhash would look like:

```
POST /_matrix/media/r0/upload?generate_blurhash=true&filename=My+Family+Photo.jpeg HTTP/1.1
Content-Type: Content-Type: image/jpeg

<bytes>
```

Example response:

```
{
  "content_uri": "mxc://example.com/abcde123",
  "blurhash": "LKO2?U%2Tw=w]~RB"
}
```

We explicitly make this behaviour opt-in as it is assumed that the majority of
clients that end up supporting BlurHashes will be capable of generating them
locally. Thus the less load we can put on the homeserver, (by not making
blurhash generation the default), the better.

Note that media servers will not be able to return a BlurHash string for
encrypted media; that must be left to the client.

The server could additionally return the BlurHash string for an image when
given an MXC URL. This would be through something like the Media Information
API (specified in
[MSC2380](https://github.com/matrix-org/matrix-doc/pull/2380)), or similar.

## Visualisation

Viewing an image message that is loading:

![blurhashed preview](images/2448-blurhash.png)

Once the image loads:

![the image has loaded](images/2448-loaded-image.png)

> For reference, the current state of things in Element is:

Reviewer 5: Slight wording change to de-emphasize Element (given it has no bearing on spec outside of ideas), and to take out time-based problems:
```suggestion
As a sample, this is Element's behaviour prior to this MSC's introduction:
```

```json
{
  "og:title": "Matrix Blog Post",
  "og:description": "This is a really cool blog post from matrix.org",
  "og:image": "mxc://example.com/ascERGshawAWawugaAcauga",
  "og:image:type": "image/png",
  "og:image:height": 48,
  "og:image:width": 48,
  "matrix:image:size": 102400,
  "matrix:image:blurhash": "oyp8ky2BWn7VHEL"
}
```

### Inline images

An optional attribute is added to `<img>` tags in messages:
`data-mx-blurhash`, where the value of the attribute is the blurhash
representation of the inline image.

This would be optionally displayed while the inline image itself is loaded in
parallel.

Example `m.room.message.formatted_body`:

```
"formatted_body": This is awesome <img alt=\"flutterjoy\" title=\"flutterjoy\" height=\"32\" src=\"mxc://matrix.example.org/abc\" data-mx-blurhash=\"LEHV6nWB2yk8pyo\" />
```

## Calculating a blurhash on the server

BlurHashes are inserted into events by the client, however some clients may not
be able to implement the BlurHash library for whatever reason. In this case, it
would be nice to allow the media repository to calculate the BlurHash of a piece
of media for the client, similar to how thumbnails are calculated by media

> repositories today.

Author: It's roughly two years on. Are there any valid use cases that we're aware of for this endpoint today? I don't know of any clients that have needed to rely on it.

Reviewer 6: No. Element sometimes relies on the media repo for thumbnails, but that doesn't work in encrypted rooms anyway and I haven't seen any client using that endpoint nor any server supporting it.

Author: Right. I'm going to rip it out of this MSC then. cc @turt2live who originally(?) proposed it.

Reviewer 5: I've been chasing memory allocation issues related to blurhashes in matrix-media-repo (MMR), and my conclusion is I agree with this conclusion.

Element is not using the returned field, and most/all of the bridges I can test quickly aren't using it either. This makes it fairly pointless for servers to support.

----

For added context, the allocation issue is with the *input* for calculating a blurhash. The blurhash itself is a non-issue.

![image](https://github.com/matrix-org/matrix-spec-proposals/assets/1190097/06d0b6d5-6818-43e8-af32-a98ea2af2369)

Specifically, jpegs are awful.

Content-Type: image/jpeg

<bytes>
```

Example response:

```
{
  "content_uri": "mxc://example.com/abcde123",
  "blurhash": "LKO2?U%2Tw=w]~RB"
}
```

We explicitly make this behaviour opt-in as it is assumed that the majority of
clients that end up supporting BlurHashes will be capable of generating them
locally. Thus, the less load we can put on the homeserver (by not making
blurhash generation the default) the better.

Note that media servers will not be able to return a BlurHash string for
encrypted media; that must be left to the client.

The server could additionally return the BlurHash string for an image when
given an MXC URL. This would be through something like the Media Information
API (specified in
[MSC2380](https://github.com/matrix-org/matrix-doc/pull/2380)), or similar.

## Visualisation

Viewing an image message that is loading:

![blurhashed preview](images/2448-blurhash.png)

Once the image loads:

![the image has loaded](images/2448-loaded-image.png)

As a sample, this is Element's behaviour prior to this MSC's introduction:

![boo, sad](images/2448-current-state.png)

## Alternatives

We could include a base64 thumbnail of the image in the event, but blurhash

> produces much more efficient textual representations.

Author: Discuss the alternative algorithm (and self-proclaimed improvement over BlurHash) [ThumbHash](https://evanw.github.io/thumbhash/).

Reviewer 6: The alpha support is nice.

Author: A comparison of ThumbHash vs. BlurHash.

## Pros

### Alpha support

ThumbHash handles images with transparent backgrounds in a much nicer fashion than BlurHash. While the obvious use case for this is stickers, as [mentioned elsewhere](https://github.com/matrix-org/matrix-spec-proposals/pull/2448#discussion_r934576480), I think that vector outlines would be the best way to represent stickers before they are downloaded in full resolution.

Still, being able to somewhat make out what an image with a transparent background is before it has loaded is valuable.

![image](https://github.com/matrix-org/matrix-spec-proposals/assets/1342360/25609354-d64b-4462-a301-733d468d2ac8)

### Better quality

ThumbHash appears to generally create a better representation of the image than BlurHash (examples taken from https://evanw.github.io/thumbhash/):

![image](https://github.com/matrix-org/matrix-spec-proposals/assets/1342360/16879a7c-794e-4335-8780-3d291e5c8e3a)

![image](https://github.com/matrix-org/matrix-spec-proposals/assets/1342360/cdca5153-ec8a-4d7d-9e3d-3082ec0efcc6)

![image](https://github.com/matrix-org/matrix-spec-proposals/assets/1342360/7df3c6b8-6a4f-4e77-9a1d-9b7aa4faf51e)

## Cons

### Limited Library Support

BlurHash was one of the first to widely publicise this use case, and thus it is a lot more popular than ThumbHash. Compare the number of implementations for [ThumbHash](https://github.com/evanw/thumbhash#implementations) versus [BlurHash](https://github.com/woltapp/blurhash#implementations).

Still, the algorithm is so simple that you could presumably translate it into your chosen language in about 30m.

### BlurHash is already widely used in Matrix

Again, due to BlurHash coming out much earlier than ThumbHash, Matrix clients have already implemented BlurHash (through this MSC) widely. If we switch, clients with concern for backwards-compatibility will likely need to implement both BlurHash and ThumbHash.

However, currently BlurHash mostly applies to media sent in the timeline, which quickly becomes stale. Element Web only supports `m.image` events in the timeline. Nheko's BlurHash support extends to both stickers and map previews on location events. I'd be most concerned about room/user avatars, which rarely change. But I've not yet seen clients implement support for that yet.

The more pressing concern would be interacting with older clients that still only send BlurHashes instead of ThumbHashes. However the failure mode here during the transition period wouldn't be too bad - you just won't see blurred thumbnails.

### Bandwidth

TODO: I'd like to conduct a test of both algorithms over a range of, say, 100 images. Using base83 encoding for both. Currently I'm not sure whether BlurHash or ThumbHash generally produces smaller encoding sizes.

Reviewer 16: ThumbHash is also around 10x faster to generate, at least on iOS.  I can run a little experiment and report the results if there's interest.

The Circles client is no longer generating BlurHashes.  We will continue to display them but we are moving entirely to ThumbHash for the future.  The slow performance was a big part of this, but also ThumbHash just seems to work better all around.  For example, we were having issues on Android where the BlurHash code crashed the app on an invalid input.

Reviewer 12: As the person who implemented Blurhash based on this MSC in Element Web I'm all for switching to something which supports alpha and has better performance, maintaining blurhash rendering similar to how @cvwright described another client handling it for some time is acceptable.

Reviewer 6: Thumbhash shouldn't be significantly faster. What you are probably seeing is that most of the thumbhash libraries downscale images to 32x32 pixels before generating the hash, while blurhash libs expect the library user to do that. If you do such an experiment, make sure you take that into account.

Reviewer 16: No, I downsample the source image once to 100x100.  Then from that 100x100 I create the ThumbHash and the BlurHash.

Maybe the Swift ThumbHash implementation is just more optimized than the BlurHash version?  The author went to some pretty great lengths to make it fast.

Reviewer 6: It doesn't look particularly optimized, but neither was the blurhash implementation, I guess. But feel free to bench it. Seems like I was also wrong, the hashing doesn't downscale first, but rendering the blurhash does create a 32x32 pixels image at best, which would be 10x faster than creating a 100x100 pixels image. So would be interesting to see comparisons, especially if you include my lib: https://github.com/Nheko-Reborn/blurhash :)

Author: @cvwright If you are doing a benchmark, reporting on the resulting filesizes of the hashes would be appreciated!

Reviewer 12: @anoadragon453 if this is to progress we probably need to update the MSC to define the field for the thumbhash to live. The lack of alpha channel in blurhash makes it insufficient in my opinion.

I have a PoC impl for EW which sends & renders both blurhash & thumbhash, preferring the latter for rendering when available.

Blurhash
![image](https://github.com/matrix-org/matrix-spec-proposals/assets/2403652/4b93c3bb-5fcc-4023-a83e-469394da2581)
Thumbhash
![image](https://github.com/matrix-org/matrix-spec-proposals/assets/2403652/a57a2131-d653-432f-b0f8-31c1f0a30490)
Original
![image](https://github.com/matrix-org/matrix-spec-proposals/assets/2403652/86bf1a1a-1bea-48e5-b3bf-e174c5c5dc9c)

PoC: `content["xyz.amorgan.thumbhash"]` base64 encoded https://github.com/matrix-org/matrix-react-sdk/pull/12164

# MSC2448: Using BlurHash in Media Events

[BlurHash](https://blurha.sh) is a compact representation of a placeholder
for an image (or the frame of a video). Currently in Matrix, clients must
display a placeholder image in the message timeline while a piece of media is
loading. Some clients, such as Riot, simply display an empty space.

While thumbnails exist to combat this to some degree, they still need to be
downloaded from a homeserver, which is not instantaneous.

Instead, a blurhash can be sent inside the `m.room.message` event, which upon
receipt other clients can render for a pretty preview while the actual

> thumbnail downloads.

Author: Not sure. Generally I don't find they ever take very long to load, but I suppose they could if you post big ones. I suppose it can't hurt really.

> thumbnail downloads.

Reviewer 17: :+1: for this, some bots and stuff kinda abuse inline lol

> ## Proposal

Reviewer 5: Would it make sense to have the media repo optionally return the hash on `/upload` too? For super lazy clients or bots that might be doing weird things with the image (ie: streaming directly from a generated buffer or something).

> A new field is added in `m.room.message`'s content field called `blurhash`.

Reviewer 4: ```suggestion
A new optional field is added in `m.room.message`'s content field called `blurhash`.
```

Reviewer 4: Also, I wonder if it's worth mentioning that blurhash is only relevant to certain `msg_type`s. It's implied in some places, but not explicitly mentioned anywhere.

It is a BlurHash of the original piece of media. Clients could then render
this using [one of the available BlurHash
implementations](https://github.com/woltapp/blurhash).

> This would be displayed while the thumbnail of the media is loaded in parallel.

Reviewer 5: Let's not force clients to render media in specific ways (thumbnails are already optional):
```suggestion
This would optionally be displayed while the thumbnail of the media is loaded in parallel.
```

* `m.image`
* `m.video`

### m.sticker

An optional field is added to `m.sticker`'s `content.info` dictionary with
the key `blurhash`. Its value is a BlurHash of the sticker media.

Example `m.sticker` content:

```json
{
  "body": "Landing",
  "info": {
    "h": 200,
    "mimetype": "image/png",
    "size": 73602,
    "thumbnail_info": {
      "h": 200,
      "mimetype": "image/png",
      "size": 73602,
      "w": 140
    },
    "thumbnail_url": "mxc://matrix.org/sHhqkFCvSkFwtmvtETOtKnLP",
    "w": 140,
    "blurhash": "JadR*.7kCMdnj"
  },
  "url": "mxc://matrix.org/sHhqkFCvSkFwtmvtETOtKnLP"
}
```

### m.room.avatar

Room avatars having BlurHashes available will be especially useful when
viewing a server's Public Rooms directory.

An optional field is added to `m.room.avatar`'s `content.info` dictionary with the
key `blurhash`. Its value is a BlurHash of the media that is pointed to by
`url`.

Example `m.room.avatar` content:

```json
{
  "url": "mxc://amorgan.xyz/a59ee02f180677d83d1b57d366127f8e1afdd4ed",
  "info": {

>     "blurhash": "JadR*.7kCMdnj"

Reviewer 18: There's a problem with avatar blurhashes - they don't have dimensions defined in the event, and the Blurhash algorithm doesn't allow to encode full dimensions, AFAICT. Meanwhile (and that's especially relevant to room avatars) dimensions may diverge from square. E.g., if I happen to use a 140x310 image I'll have to either: 1) crop/resize it to something with a more reasonable ratio (140x300 or even 140x280); 2) accept that my blurhash will have 14x31 components; 3) accept that the dimensions may change when cross-fading to the real image. Moreover, Quaternion, e.g., adjusts the UI to the room avatar dimensions. Having a blurhash that only has a ratio but not dimensions doesn't help and leaves me with option 3. While for user avatars that might be an overkill (non-square user avatars are not good for pills etc.), it would be good to also have `w` and `h` added to `info` for room avatars.

Reviewer 6: While I do agree, that this will cause an issue, maybe some explanation around the height and width components of a blurhash: A blurhash only can have up to 9 components in either direction and they don't really matter when decoding apart from the quality being lower, if you have fewer components. As such it is completely fine for a client to encode all images with 3x4 components or similar and users will barely notice that and a client can't rely on the aspect ratio of the blurhash matching the aspect ratio of the image.

So the issue here is decoding and it doesn't significantly worsen the current situation:
- If your UI adapts to the aspect ratio of the room or user avatar, it would already need to resize whatever placeholder it showed before, because avatars lack the dimension information.
- Because you lack dimension information, you may decode the image to the size of your placeholder, which does not match the dimension of the final crop or fit, as such the image would maybe look significantly different than the blurhash looked, if the image was extensively long or wide.

I would appreciate in general having an aspect ratio/size for an avatar though, since this was already somewhat of an issue before, if your client wanted to show a better placeholder and blurhashes make the issue worse, but I would say this is a preexisting spec issue. Also non square user avatars are in general a pain and clients maybe should just prompt the user to resize them to be square to make them look similar across clients.

Reviewer 18: Thanks, I missed on the 9x9 limitation; as I said, it's room avatars that I'm mostly concerned about since there are more options to display them non-rectangular and larger than one or two lines height in the room header. I agree that adding blurhashes without adding dimensions doesn't worsen the current situation but as we are to spec that I'd rather do it the right way straight away. In theory it could be a separate MSC but I doubt it's worth it to slice that thin.

Reviewer 6: Yep, I agree, having dimensions in the Avatar is nice and could be lumped in here.

Author: Sorry, I failed to include it in the example for this MSC, but we do already have height and width dimensions for `m.room.avatar` events defined in the spec: https://matrix.org/docs/spec/client_server/r0.6.1#m-room-avatar

I believe the reason I failed to do so is that it looks like Element Web (at least) doesn't populate that information. So the real solution (other than perhaps making the values required, which would be hard to do retroactively) may just be to expand client support for the optional values?

Reviewer 18: I should have looked in the spec before asking... Yes it would be great to have it actually supported by Element clients; and no changes to the MSC are needed then.

Reviewer 12: Do any other clients support it? Anoa and I checked our HS databases and found no legitimate uses of the field in room avatars our servers have seen. There were invalid usages however.

![image](https://user-images.githubusercontent.com/2403652/119275084-fd484980-bc0a-11eb-999b-a79a6e983096.png)

}
```

A server MUST also be able to render a BlurHash image from a BlurHash string.
This is to support clients that do not have a BlurHash client implementation
or are otherwise unable to do so themselves.

This takes the form of a new endpoint, `GET
/_matrix/media/r0/blurhash/{encodedBlurHash}`. This endpoint supports the following query parameters:

* `width` - The width in pixels of the returned BlurHash image
* `height` - The height in pixels of the returned BlurHash image

Note that implementations should be careful to limit the size of accepted
BlurHash strings, as not to overload the server with processing an obscenely
long string. If this is the case, the server should return a `400
M_TOO_LARGE`.

Example request:

```
GET /_matrix/media/r0/blurhash/LG.F5%5D%3B%2BYk%5E6%23%25*%2B%2CK-%3A9%3D%3F%40%5B

> {}

Reviewer 5: surely we don't need a request body?

Author: Oh, this was to signify that you don't, but indeed it may signify the opposite to the user.

Will remove.

}
```

> Servers will obviously not be able to return a BlurHash string for encrypted
> media.

Reviewer 18: When reading the first time, I didn't recognise it only pertains to `/upload` responses and not to (home)servers sending encrypted attachment events - hence suggesting.
```suggestion
Media servers will obviously not be able to return a BlurHash string for encrypted
media; clients have no other way but to calculate it on their side.
```
